external/fv3fit/fv3fit/pytorch/predict.py

```diff
@@ -1,76 +1,128 @@
+from .._shared.scaler import StandardScaler
 import numpy as np
 import torch
 import torch.nn as nn
 import xarray as xr
-from fv3fit._shared import (
-    Predictor,
-    match_prediction_to_input_coords,
-    SAMPLE_DIM_NAME,
-    stack,
-)
-from typing import Any, Dict, Hashable, Iterable, Sequence
+from fv3fit._shared import Dumpable
+from typing import Any, Dict, Hashable, Iterable, Mapping, Tuple
+from fv3fit.pytorch.system import DEVICE
 
 
-class PytorchModel(Predictor):
+class PytorchModel(Dumpable):
 
     _MODEL_FILENAME = "weight.pt"
     _CONFIG_FILENAME = "config.yaml"
     custom_objects: Dict[str, Any] = {}
 
     def __init__(
         self,
-        input_variables: Iterable[Hashable],
-        output_variables: Iterable[Hashable],
+        state_variables: Iterable[Hashable],
         model: nn.Module,
-        unstacked_dims: Sequence[str],
+        scalers: Mapping[Hashable, StandardScaler],
     ):
         """Initialize the predictor
         Args:
-            input_variables: names of input variables
-            output_variables: names of output variables
+            state_variables: names of state variables
             model: pytorch model to wrap
-            unstacked_dims: non-sample dimensions of model output
+            scalers:
         """
-        super().__init__(input_variables, output_variables)
-        self.input_variables = input_variables
-        self.output_variables = output_variables
+        self.state_variables = state_variables
         self.model = model
-        self._unstacked_dims = unstacked_dims
+        self.scalers = scalers
 
-    def _array_prediction_to_dataset(
-        self, names, outputs, stacked_coords
-    ) -> xr.Dataset:
-        ds = xr.Dataset()
-        for name, output in zip(names, outputs):
-            dims = [SAMPLE_DIM_NAME] + list(self._unstacked_dims)
-            scalar_singleton_dim = (
-                len(output.shape) == len(dims) and output.shape[-1] == 1
-            )
-            if scalar_singleton_dim:  # remove singleton dimension
-                output = output[..., 0]
-                dims = dims[:-1]
-            da = xr.DataArray(
-                data=output, dims=dims, coords={SAMPLE_DIM_NAME: stacked_coords},
-            ).unstack(SAMPLE_DIM_NAME)
-            dim_order = [dim for dim in self._unstacked_dims if dim in da.dims]
-            ds[name] = da.transpose(*dim_order, ...)
-        return ds
+    def pack_to_tensor(self, ds: xr.Dataset, times_per_window: int) -> torch.Tensor:
```

Comment:
The way I wrote the packing, the times in the window don't overlap, which means if you divide the dataset evenly into 1-day windows there will be one time of day where we never predict (the end of one window, which would evolve to the start of the next window). Is this acceptable or should I refactor it to pack into windows whose start and end times overlap with one another?

Reply:
Does this drastically reduce the available training data?  As long as we don't have a systematic time of day that is missing from optimization, I don't necessarily see a problem w/ non-overlapping windows.

Reply:
This is only meant to be used for evaluation. The way I made the training data loader, it will randomly select windows like bootstrapping, so each epoch some samples may be over-sampled or not sampled.